## Otter Interview Project - Analysis on restaurants with delivery services in the US ##

## 1. Initial Setup ##

#### 1.1 Library required for the analysis ####

In [547]:
library(tidyr)
library(dplyr)
library(ggmap)
library(psych)
library(maps)
library(ggplot2)
library(xray)
library(data.table)
library(finalfit)
library(tidygeocoder)
library(summarytools)

#### 1.2 Data Preparation ####

- Set the working directory
- Read the file
- Ensure the dataset is prepared and converted into proper dataframe format

In [548]:
## setwd("/Users/tonychu/SynologyDrive/Resume:Application related document/Otter")
df <- read.table("css_public_all_ofos_locations.csv",header=TRUE, sep='\001',na.strings ='\\N', quote = "", fill=T)

In [549]:
head(df)
dim(df)

restaurant_id name                                      platform sub_platform
1 9744          Loco Coco                                 caviar               
2 4903          Musubi                                    caviar               
3 4723          The Bakers' Lounge                        caviar               
4 4128          Parm - Battery Park City                  caviar               
5 5690          Derek Test Merchant                       caviar               
6 2827          The Plant Cafe - California & Pine (FiDi) caviar               
  latitude  longitude   city country active
1 40.765004 -73.965961       US      TRUE  
2 45.504641 -122.643806      US      FALSE 
3 38.924559 -76.990115       US      FALSE 
4 40.71418  -74.015568       US      TRUE  
5 1         2                US      FALSE 
6 37.792829 -122.397881      US      FALSE 
  standardized_name                     restaurant_chain delivery_radius geom
1 loco_coco                                                              NA  
2 musubi                                                                 NA  
3 the_bakers_lounge                                                      NA  
4 parm_-_battery_park_city                                               NA  
5 derek_test_merchant                                                    NA  
6 the_plant_cafe_-_california_pine_fidi                                  NA

[1] 644089     13

##### 1.3 Convert the attributes into the proper data type #####

- Change the data type of **restaurant_id, latitude, and longitude** into numeric variables

In [550]:
num_col = c("restaurant_id","latitude", "longitude","delivery_radius")
df[ , num_col] <- apply(df[ , num_col], 2, function(x) as.numeric(as.character(x)))
head(df)

Warning message in FUN(newX[, i], ...):
“NAs introduced by coercion”
Warning message in FUN(newX[, i], ...):
“NAs introduced by coercion”
Warning message in FUN(newX[, i], ...):
“NAs introduced by coercion”
Warning message in FUN(newX[, i], ...):
“NAs introduced by coercion”


restaurant_id name                                      platform sub_platform
1 9744          Loco Coco                                 caviar               
2 4903          Musubi                                    caviar               
3 4723          The Bakers' Lounge                        caviar               
4 4128          Parm - Battery Park City                  caviar               
5 5690          Derek Test Merchant                       caviar               
6 2827          The Plant Cafe - California & Pine (FiDi) caviar               
  latitude longitude  city country active standardized_name                    
1 40.76500  -73.96596      US      TRUE   loco_coco                            
2 45.50464 -122.64381      US      FALSE  musubi                               
3 38.92456  -76.99012      US      FALSE  the_bakers_lounge                    
4 40.71418  -74.01557      US      TRUE   parm_-_battery_park_city             
5  1.00000    2.00000      US      FALSE  derek_test_merchant                  
6 37.79283 -122.39788      US      FALSE  the_plant_cafe_-_california_pine_fidi
  restaurant_chain delivery_radius geom
1                  NA              NA  
2                  NA              NA  
3                  NA              NA  
4                  NA              NA  
5                  NA              NA  
6                  NA              NA

## 2. Data Exploration and data cleaning ##
    

- Look at the summary of the dataframe consisting of: variable names and labels, factor levels, frequencies or numerical summary statistics, and valid/missing observations information.
- There are 644,089 entries in total
- Seems like plenty of the attributes require further examinations, and below are some of the questions need to be addressed regarding data exploration:
    - Is there any missing restaurant ID, and would missing restaurant ID impact the analysis?
    - How many delivery platforms and sub platforms are there?
    - Are there restaurants with missing coordinates? Can I identify duplicated restaurants with no coordinates?
    - How many restaurants are missing the city attribute? How do I extract city information without city attribute? Reverse check-up using the coordinates?
    - Are we certain all restaurants are all located in the US?
    - How do I standardize restaurant names and how do I remove duplicated entries 

In [551]:
subset(dfSummary(df), select=-Graph)

No Variable                        
1   1 restaurant_id\\\n[numeric]      
2   2 name\\\n[character]             
3   3 platform\\\n[character]         
4   4 sub_platform\\\n[character]     
5   5 latitude\\\n[numeric]           
6   6 longitude\\\n[numeric]          
7   7 city\\\n[character]             
8   8 country\\\n[character]          
9   9 active\\\n[character]           
10 10 standardized_name\\\n[character]
11 11 restaurant_chain\\\n[character] 
12 12 delivery_radius\\\n[numeric]    
13 13 geom\\\n[logical]               
   Stats / Values                                                                                                                                                                                                                                                            
1  Mean (sd) : 237866.8 (187782.5)\\\nmin < med < max:\\\n1 < 216563 < 997804\\\nIQR (CV) : 249348 (0.8)                                                                                                                                                                     
2  1\\. Subway\\\n2\\. Starbucks\\\n3\\. McDonald's\\\n4\\. Taco Bell\\\n5\\. Burger King\\\n6\\. 7-Eleven\\\n7\\. KFC\\\n8\\. Walgreens\\\n9\\. Dunkin' Donuts\\\n10\\. Wendy's\\\n[ 418989 others ]                                                                        
3  1\\. doordash\\\n2\\. postmates\\\n3\\. grubhub\\\n4\\. ubereats\\\n5\\. delivery.com\\\n6\\. caviar\\\n7\\. (Empty string)\\\n8\\. ·[""open: 12:00\\\n9\\. ·[]\\\n10\\. ·[""open: 11:00\\\n[ 37 others ]                                                                 
4  1\\. doordash\\\n2\\. restaurant\\\n3\\. grubhub\\\n4\\. ubereats\\\n5\\. (Empty string)\\\n6\\. ·close: 21:00""]\\\n7\\. ·close: 20:00""]\\\n8\\. ·[]\\\n9\\. ·close: 22:00""]\\\n10\\. ·close: 00:00""]\\\n[ 21 others ]                                                
5  Mean (sd) : 37 (8.5)\\\nmin < med < max:\\\n-90 < 38.5 < 555\\\nIQR (CV) : 7 (0.2)                                                                                                                                                                                        
6  Mean (sd) : -94 (19.1)\\\nmin < med < max:\\\n-158.3 < -87.8 < 555\\\nIQR (CV) : 37.7 (-0.2)                                                                                                                                                                              
7  1\\. (Empty string)\\\n2\\. Los Angeles\\\n3\\. New York\\\n4\\. Houston\\\n5\\. Chicago\\\n6\\. Miami\\\n7\\. Philadelphia\\\n8\\. San Diego\\\n9\\. San Francisco\\\n10\\. Orlando\\\n[ 3050 others ]                                                                   
8  1\\. US\\\n2\\. (Empty string)\\\n3\\. ·close: 21:00""]\\\n4\\. ·close: 20:00""]\\\n5\\. FALSE\\\n6\\. ·close: 22:00""]\\\n7\\. ·close: 00:00""]\\\n8\\. ·close: 20:30""]\\\n9\\. ·close: 23:00""]\\\n10\\. ·close: 21:30""]\\\n[ 18 others ]                             
9  1\\. TRUE\\\n2\\. FALSE\\\n3\\. (Empty string)\\\n4\\. ·close: 21:00""]\\\n5\\. ·close: 20:00""]\\\n6\\. ·close: 22:00""]\\\n7\\. ·close: 00:00""]\\\n8\\. ·close: 23:00""]\\\n9\\. ·[]]"\\\n10\\. 40\\\n[ 21 others ]                                                    
10 1\\. subway\\\n2\\. starbucks\\\n3\\. mcdonald_s\\\n4\\. taco_bell\\\n5\\. burger_king\\\n6\\. jack_in_the_box\\\n7\\. 7-eleven\\\n8\\. kfc\\\n9\\. dunkin_donuts\\\n10\\. walgreens\\\n[ 409282 others ]                                                                 
11 1\\. (Empty string)\\\n2\\. TRUE\\\n3\\. ·close: 21:00""]]"\\\n4\\. ·close: 20:00""]]"\\\n5\\. ·close: 00:00""]]"\\\n6\\. ·close: 22:00""]]"\\\n7\\. ·close: 23:00""]]"\\\n8\\. ·close: 21:30""]]"\\\n9\\. ·close: 18:00""]]"\\\n10\\. ·close: 19:00""]]"\\\n[ 26 others ]
12 Mean (sd) : 43787.3 (22251810)\\\nmin < med < max:\\\n1 < 7000 < 13235083092\\\nIQR (CV) : 1500 (508.2)                                                                                                               

##### 2.1 Data Cleaning by restaurant ID #####

- By examining restaurants with no restaurant ID, seems like i am still able to find out the information that I required without restaurant ID.

In [552]:
head(subset(df,is.na(restaurant_id)),20)

restaurant_id name                    platform     sub_platform latitude
34795 NA                                    delivery.com                    NA
94112 NA            Hapa Sushi LoDo         postmates    restaurant   39.74985
94113 NA            Fat Sal's               postmates    restaurant   34.06253
94114 NA            Monk's Kettle           postmates    restaurant   37.76489
94115 NA            Firehouse Subs          postmates    restaurant   41.70156
94116 NA            Carl's Jr/Green Burrito postmates    restaurant   33.81533
94117 NA            Dunkin' Donuts          postmates    restaurant   40.71571
94118 NA            Subway                  postmates    restaurant   39.81755
94119 NA            7-Eleven                postmates    restaurant   38.96194
94120 NA            Anson Eleven            postmates    restaurant   31.75912
94121 NA            Longhorn Steakhouse     postmates    restaurant   31.52358
94122 NA            Smashburger             postmates    restaurant   33.57749
94123 NA            Sonic Drive-In          postmates    restaurant   33.62390
94124 NA            Walgreens               postmates    restaurant   29.64925
94125 NA            Bangkok Grill           postmates    restaurant   34.11146
94126 NA            Starbucks               postmates    restaurant   37.71328
94127 NA            Chipotle Mexican Grill  postmates    restaurant   36.78203
94128 NA            Whole Foods Market      postmates    restaurant   34.75047
94129 NA            Whole Foods Market      postmates    restaurant   38.99610
94130 NA            Whole Foods Market      postmates    restaurant   38.90108
      longitude  city country active standardized_name       restaurant_chain
34795         NA      US      FALSE                                          
94112 -104.99994      US      TRUE   hapa_sushi_lodo                         
94113 -118.44805      US      TRUE   fat_sal_s                               
94114 -122.42291      US      TRUE   monk_s_kettle                           
94115  -83.65030      US      TRUE   firehouse_subs                          
94116 -116.40101      US      FALSE  carl_s_jr_green_burrito                 
94117  -73.59847      US      TRUE   dunkin_donuts                           
94118  -86.32782      US      TRUE   subway                                  
94119  -76.86317      US      FALSE  7-eleven                                
94120 -106.48910      US      TRUE   anson_eleven                            
94121  -97.16627      US      FALSE  longhorn_steakhouse                     
94122 -117.72478      US      TRUE   smashburger                             
94123 -112.42860      US      TRUE   sonic_drive-in                          
94124  -95.14302      US      FALSE  walgreens                               
94125 -118.24986      US      TRUE   bangkok_grill                           
94126 -120.95271      US      TRUE   starbucks                               
94127 -119.78952      US      TRUE   chipotle_mexican_grill                  
94128  -92.40598      US      FALSE  whole_foods_market                      
94129  -77.02450      US      FALSE  whole_foods_market                      
94130  -77.04948      US      FALSE  whole_foods_market                      
      delivery_radius geom
34795 NA              NA  
94112 NA              NA  
94113 NA              NA  
94114 NA              NA  
94115 NA              NA  
94116 NA              NA  
94117 NA              NA  
94118 NA              NA  
94119 NA              NA  
94120 NA              NA  
94121 NA              NA  
94122 NA              NA  
94123 NA              NA  
94124 NA              NA  
94125 NA              NA  
94126 NA              NA  
94127 NA              NA  
94128 NA              NA  
94129 NA              NA  
94130 NA              NA

##### 2.2 Data cleaning by restaurant name #####

- Change all restaurant names to lower case to simplify data deduplication
- Let's examine the name of the restaurant. Apparently, only 65 restaurants have no proper name entry nor other information. So they can comfortably be removed as it only made up small portion of the overall dataset.

In [553]:
df$name <- tolower(df$name)

- Create a subset of entries with no restaurant names

In [564]:
head(subset(df, name==""))
no_name <- subset(df, name=="")
dim(no_name)

restaurant_id name platform     sub_platform latitude longitude city
675    4075                                        NA       NA            
27428 64030                                        NA       NA            
29810   646                                        NA       NA            
29929 63819                                        NA       NA            
34795    NA              delivery.com              NA       NA            
41511 72462                                        NA       NA            
      country active standardized_name restaurant_chain delivery_radius geom
675                                                     NA              NA  
27428                                                   NA              NA  
29810                                                   NA              NA  
29929                                                   NA              NA  
34795 US      FALSE                                     NA              NA  
41511                                                   NA              NA

[1] 65 13

- Assuming due to system error or manual entries, 170 of the entries have information recorded that cannot be contributed to the analysis as well. So we will remove them also.
- By further examining the name attribute, some entries have "duplicate","training","do not use","do not enable","churned","closed" in their names, thus I automatically assume this is mainly due to manual entry or name given by system due to specific reasons. There are 5,839 entries like this, and I decide to remove them because they also only made up a very small portion of the dataset. 

In [565]:
head(filter(df, grepl('duplicate|training|do not use|do not enable|churned|closed|close:|open:|usd', name)),10)
bad_entry <- filter(df, grepl('duplicate|training|do not use|do not enable|churned|closed|close:|open:|usd', name))
dim(bad_entry)

restaurant_id name                                                  
1   6878         chen's chinese & cuisine {do not use}                 
2   8216         king��s road cafe - do not enable                     
3  73588         andrew's diner - using for training                   
4  88006         oob - closed - lee chi                                
5   5074         yucatan taco stand- do not enable                     
6   3700         [do not enable, bad duplicate] meehan's (sandy sprigs)
7    601         tiffin bistro duplicate                               
8    803         pure fare - south street duplicate                    
9    725         phuong thao [do not enable]                           
10  5309         hot wok bistro too (do not use)                       
   platform     sub_platform latitude longitude city country active
1  caviar                    41.8889   -87.6330      US      FALSE 
2  caviar                    34.0762  -118.3723      US      FALSE 
3  delivery.com              40.5383   -74.1490      US      FALSE 
4  delivery.com              40.7769   -73.9525      US      FALSE 
5  caviar                    32.8153   -96.7703      US      FALSE 
6  caviar                    33.8395   -84.3771      US      FALSE 
7  caviar                    39.9347   -75.1622      US      FALSE 
8  caviar                    39.9442   -75.1692      US      FALSE 
9  caviar                    37.3769  -122.0305      US      FALSE 
10 caviar                    37.4958  -122.2479      US      FALSE 
   standardized_name                                 restaurant_chain
1  chen_s_chinese_cuisine_do_not_use                                 
2  king_s_road_cafe_-_do_not_enable                                  
3  andrew_s_diner_-_using_for_training                               
4  oob_-_closed_-_lee_chi                                            
5  yucatan_taco_stand-_do_not_enable                                 
6  do_not_enable_bad_duplicate_meehan_s_sandy_sprigs                 
7  tiffin_bistro_duplicate                                           
8  pure_fare_-_south_street_duplicate                                
9  phuong_thao_do_not_enable                                         
10 hot_wok_bistro_too_do_not_use                                     
   delivery_radius geom
1  NA              NA  
2  NA              NA  
3  NA              NA  
4  NA              NA  
5  NA              NA  
6  NA              NA  
7  NA              NA  
8  NA              NA  
9  NA              NA  
10 NA              NA

[1] 6009   13

##### 2.3 Data cleaning by coordinates #####

- There are 2,042 entries with no coordinates available
- There are 1,759 entries with wrong coordinates that point to either the North Pole, South Pole, or other extreme locations
- So it is safe to say that we can remove these entries as we have no ways to find out where their exact locations are
- Coordinates should be round up the 4th decimal place since each degree in longitude and latitude is around 111 KM. It is reasonable to assume that there can be different restaurants 10 meter apart on the map as 10 meter equates to 0.0001 degree in terms of coordinates.

In [566]:
df$latitude <- round(df$latitude,4)
df$longitude <- round(df$longitude,4)

In [567]:
head(subset(df, is.na(longitude) & is.na(latitude)),10)
no_cor <- subset(df, is.na(longitude) & is.na(latitude))
dim(no_cor)

restaurant_id name                        platform sub_platform latitude
675   4075                                                           NA      
1927 86915         el sazon de las americas                          NA      
2092  8142         ten sushi                                         NA      
4167 54962         al's                                              NA      
4387 61610         fake restaurant - reviewer                        NA      
4950 87158         mocha's cafe & grill                              NA      
5908 56614         chili's grill & bar                               NA      
6005 30732         extreme pizza: template (                         NA      
6218 35120         soho thai                                         NA      
6388 93074         jasim's                                           NA      
     longitude city country active standardized_name restaurant_chain
675  NA                                                              
1927 NA                                                              
2092 NA                                                              
4167 NA                                                              
4387 NA                                                              
4950 NA                                                              
5908 NA                                                              
6005 NA                                                              
6218 NA                                                              
6388 NA                                                              
     delivery_radius geom
675  NA              NA  
1927 NA              NA  
2092 NA              NA  
4167 NA              NA  
4387 NA              NA  
4950 NA              NA  
5908 NA              NA  
6005 NA              NA  
6218 NA              NA  
6388 NA              NA

[1] 2042   13

In [568]:
head(subset(df, ((abs(latitude) == 90 & longitude ==0 ) | (abs(latitude) <= 1 | abs(longitude) <= 1 ) | (abs(latitude) > 90 | abs(longitude) > 180))),10)
wrong_cor <- subset(df, ((abs(latitude) == 90 & longitude ==0 ) | (abs(latitude) <= 1 | abs(longitude) <= 1 ) | (abs(latitude) > 90 | abs(longitude) > 180)))
dim(wrong_cor)

restaurant_id name                                     platform    
5    5690         derek test merchant                      caviar      
61  59739         bequ juice                               delivery.com
65  63820         gino's east                              delivery.com
101 64603         sori sushi                               delivery.com
138  5648         jessica hui placeholder                  caviar      
161  1526         the halal guys                           caviar      
167  9812         the french bakery - 112th ave [disabled] caviar      
202 63305         ezras food                               delivery.com
213 57676         cupping room cafe                        delivery.com
218 61832         grapevine travelers                      delivery.com
    sub_platform latitude longitude city country active
5                  1       2             US      FALSE 
61               -90       0             US      FALSE 
65               -90       0             US      FALSE 
101              -90       0             US      FALSE 
138                0       0             US      FALSE 
161                0       0             US      FALSE 
167                1      -1             US      FALSE 
202              -90       0             US      FALSE 
213              -90       0             US      FALSE 
218              -90       0             US      FALSE 
    standardized_name                      restaurant_chain delivery_radius
5   derek_test_merchant                                     NA             
61  bequ_juice                                              NA             
65  gino_s_east                                             NA             
101 sori_sushi                                              NA             
138 jessica_hui_placeholder                                 NA             
161 the_halal_guys                                          NA             
167 the_french_bakery_-_112th_ave_disabled                  NA             
202 ezras_food                                              NA             
213 cupping_room_cafe                                       NA             
218 grapevine_travelers                                     NA             
    geom
5   NA  
61  NA  
65  NA  
101 NA  
138 NA  
161 NA  
167 NA  
202 NA  
213 NA  
218 NA

[1] 1763   13

##### 2.4 Data cleaning by platform status #####

- There are 2,065 entries with no delivery platform status available


In [569]:
head(subset(df, active == ""))
null_active <- subset(df, active == "")
dim(null_active)

restaurant_id name                        platform sub_platform latitude
675   4075                                                           NA      
1927 86915         el sazon de las americas                          NA      
2092  8142         ten sushi                                         NA      
4167 54962         al's                                              NA      
4387 61610         fake restaurant - reviewer                        NA      
4950 87158         mocha's cafe & grill                              NA      
     longitude city country active standardized_name restaurant_chain
675  NA                                                              
1927 NA                                                              
2092 NA                                                              
4167 NA                                                              
4387 NA                                                              
4950 NA                                                              
     delivery_radius geom
675  NA              NA  
1927 NA              NA  
2092 NA              NA  
4167 NA              NA  
4387 NA              NA  
4950 NA              NA

[1] 2065   13

##### 2.5 Data cleaning by removing unneeded subsets #####

- After removing the subsets, there are still 639,998 entries left

In [ ]:
no_name <- subset(df, name=="")
bad_entry <- filter(df, grepl('duplicate|training|do not use|do not enable|churned|closed|close:|open:|usd', name))
no_cor <- subset(df, is.na(longitude) & is.na(latitude))
wrong_cor <- subset(df, ((abs(latitude) == 90 & longitude ==0 ) | (abs(latitude) <= 1 | abs(longitude) <= 1 ) | (abs(latitude) > 90 | abs(longitude) > 180)))
null_active <- subset(df, active == "")

In [570]:
df_in_work <- df
df_in_work <- anti_join(df_in_work, no_name)
df_in_work <- anti_join(df_in_work, bad_entry)
df_in_work <- anti_join(df_in_work, no_cor)
df_in_work <- anti_join(df_in_work, wrong_cor)
df_in_work <- anti_join(df_in_work, null_active)

dim(df_in_work)

Joining, by = c("restaurant_id", "name", "platform", "sub_platform", "latitude", "longitude", "city", "country", "active", "standardized_name", "restaurant_chain", "delivery_radius", "geom")

Joining, by = c("restaurant_id", "name", "platform", "sub_platform", "latitude", "longitude", "city", "country", "active", "standardized_name", "restaurant_chain", "delivery_radius", "geom")

Joining, by = c("restaurant_id", "name", "platform", "sub_platform", "latitude", "longitude", "city", "country", "active", "standardized_name", "restaurant_chain", "delivery_radius", "geom")

Joining, by = c("restaurant_id", "name", "platform", "sub_platform", "latitude", "longitude", "city", "country", "active", "standardized_name", "restaurant_chain", "delivery_radius", "geom")

Joining, by = c("restaurant_id", "name", "platform", "sub_platform", "latitude", "longitude", "city", "country", "active", "standardized_name", "restaurant_chain", "delivery_radius", "geom")

Joining, by = c("restaurant_id", "name",

[1] 634255     13

## 3. Feature Engineering ##

##### 3.1 Restaurant name exploration and cleaning #####

- Separate names with branch names in parenthesis into two columns "restaurant_chain" & "branch"

In [572]:
df_in_work$name <- gsub("\\(|\\[", "- ", df_in_work$name)
df_in_work$name <- gsub("\\)|\\]|\\�", "", df_in_work$name)


In [573]:
df_in_work <- separate(df_in_work,name,into = c("restaurant_chain","branch"), sep=" - ", extra="merge")

Warning message:
“Expected 2 pieces. Missing pieces filled with `NA` in 430929 rows [1, 2, 3, 6, 8, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 22, 23, 27, 28, 30, ...].”


In [574]:
head(df_in_work)
dim(df_in_work)

restaurant_id restaurant_chain   branch                   platform
1 9744          loco coco          NA                       caviar  
2 4903          musubi             NA                       caviar  
3 4723          the bakers' lounge NA                       caviar  
4 4128          parm               battery park city        caviar  
5 2827          the plant cafe     california & pine - fidi caviar  
6 4850          minton's           NA                       caviar  
  sub_platform latitude longitude city country active
1              40.7650   -73.9660      US      TRUE  
2              45.5046  -122.6438      US      FALSE 
3              38.9246   -76.9901      US      FALSE 
4              40.7142   -74.0156      US      TRUE  
5              37.7928  -122.3979      US      FALSE 
6              40.8047   -73.9523      US      FALSE 
  standardized_name                     delivery_radius geom
1 loco_coco                             NA              NA  
2 musubi                                NA              NA  
3 the_bakers_lounge                     NA              NA  
4 parm_-_battery_park_city              NA              NA  
5 the_plant_cafe_-_california_pine_fidi NA              NA  
6 minton_s                              NA              NA

[1] 634255     13

- Create a list of franchise restaurant by grouping restaurant names with over 100 branches by frequencies to find unique restaurant names
- Assuming restaurants with less than 100 stores are not considered as franchises
- Run a for loop and rename restaurants that have similar names to make sure the list of restaurant names are more uniform
- Create another column type, and name those with over 100 appearances "Franchises", and the ones with less than 100 stores "SMBs"

In [575]:
franchises <- group_by(df_in_work, restaurant_chain) %>% summarize(n = n()) %>% arrange(desc(n)) %>% subset(n>100)
df_in_work$type <- ""
for (name in franchises$restaurant_chain)
{
    df_in_work$restaurant_chain[grepl(name,df_in_work$restaurant_chain)] <- name
    df_in_work$type[grepl(name,df_in_work$restaurant_chain)] <- "Franchises"
}

df_in_work$type[df_in_work$type==""] <- "SMBs"

##### 3.2 Extract location data from coordinates #####

- Using map.where() function to determine state and city information of the US based on coordinates
- Remove the ones without output as they are not locations in the United States
- the output of the function gives us state,city information, so we further split them into "state" and "city" columns

In [582]:
df_in_work$city <- map.where(database="county", df_in_work$longitude, df_in_work$latitude)
df_in_work <- subset(df_in_work, !is.na(city))
df_in_work <- separate(df_in_work,city,into = c("state","city"), sep=",")
head(df_in_work,20)

restaurant_id restaurant_chain            branch                   platform
1  9744          loco coco                   NA                       caviar  
2  4903          musubi                      NA                       caviar  
3  4723          the bakers' lounge          NA                       caviar  
4  4128          parm                        battery park city        caviar  
5  2827          the plant cafe              california & pine - fidi caviar  
6  4850          minton's                    NA                       caviar  
7  6301          brazilian bowl              albany park              caviar  
8  9006          gourmet palace china bistro NA                       caviar  
9  6208          liberty burger              keller                   caviar  
10 4972          mel's deli                  NA                       caviar  
11 5917          the office                  NA                       caviar  
12 8713          kizuki ramen & izakaya      crossroads               caviar  
13  145          4505 burgers and bbq        NA                       caviar  
14 1487          cibo                        NA                       caviar  
15  251          aperto                      NA                       caviar  
16 1458          ziti's                      NA                       caviar  
17 2269          miyako hollywood            NA                       caviar  
19 1286          doc ricketts                NA                       caviar  
20 6567          bridgetown bagels           NA                       caviar  
21 5481          coalfire pizza              west town                caviar  
   sub_platform latitude longitude state                city          country
1               40.7650   -73.9660 new york             new york      US     
2               45.5046  -122.6438 oregon               multnomah     US     
3               38.9246   -76.9901 district of columbia washington    US     
4               40.7142   -74.0156 new york             new york      US     
5               37.7928  -122.3979 california           san francisco US     
6               40.8047   -73.9523 new york             new york      US     
7               41.9686   -87.7087 illinois             cook          US     
8               34.1645  -118.4143 california           los angeles   US     
9               32.9690   -96.8203 texas                dallas        US     
10              34.0457  -118.2505 california           los angeles   US     
11              37.4941  -122.2454 california           san mateo     US     
12              47.6096  -122.1417 washington           king          US     
13              37.7762  -122.4383 california           san francisco US     
14              45.5048  -122.6274 oregon               multnomah     US     
15              37.7627  -122.3972 california           san francisco US     
16              42.3557   -71.0626 massachusetts        suffolk       US     
17              34.0981  -118.3220 california           los angeles   US     
19              37.7968  -122.4048 california           san francisco US     
20              45.5238  -122.6600 oregon               multnomah     US     
21              41.8908   -87.6599 illinois             cook          US     
   active standardized_name                     delivery_radius geom type
1  TRUE   loco_coco                             NA              NA   SMBs
2  FALSE  musubi                                NA              NA   SMBs
3  FALSE  the_bakers_lounge                     NA              NA   SMBs
4  TRUE   parm_-_battery_park_city              NA              NA   SMBs
5  FALSE  the_plant_cafe_-_california_pine_fidi NA              NA   SMBs
6  FALSE  minton_s                              NA              NA   SMBs
7  FALSE  brazilian_bowl_-_albany_park          NA              NA   SMBs
8  TRUE   gourmet_palace_china_bistro           NA              NA   SMBs
9  TRUE   liberty_burger_-_keller       

### 4. Re-arrange Dataset & Perform Deduplication

In [595]:
df_clean <- df_in_work

In [596]:
filter(df_clean, (platform != sub_platform) & (sub_platform != "restaurant") & sub_platform != "")

restaurant_id restaurant_chain branch platform sub_platform latitude
     longitude state city country active standardized_name delivery_radius geom
     type

In [621]:
subset(df_clean, is.na(platform) | platform == "")

restaurant_id restaurant_chain branch platform sub_platform latitude
     longitude state city country active standardized_name delivery_radius geom
     type

##### 4.1 Remove unwanted columns and re-arrange column orders #####

In [597]:
df_final <- select(df_clean, -c("restaurant_id","geom","sub_platform", "standardized_name"))
col_order <- c('restaurant_chain','branch','type','active','platform','latitude','longitude','city','state','country','delivery_radius')
df_final <- df_final[,col_order]
head(df_final)

restaurant_chain   branch                   type active platform latitude
1 loco coco          NA                       SMBs TRUE   caviar   40.7650 
2 musubi             NA                       SMBs FALSE  caviar   45.5046 
3 the bakers' lounge NA                       SMBs FALSE  caviar   38.9246 
4 parm               battery park city        SMBs TRUE   caviar   40.7142 
5 the plant cafe     california & pine - fidi SMBs FALSE  caviar   37.7928 
6 minton's           NA                       SMBs FALSE  caviar   40.8047 
  longitude city          state                country delivery_radius
1  -73.9660 new york      new york             US      NA             
2 -122.6438 multnomah     oregon               US      NA             
3  -76.9901 washington    district of columbia US      NA             
4  -74.0156 new york      new york             US      NA             
5 -122.3979 san francisco california           US      NA             
6  -73.9523 new york      new york             US      NA

##### 4.2 Perform deduplication #####

- To ensure each entry is unique, unique() function is required, and certain criteria need to be met:
    - A restaurant can have multiple delivery platform, despite them being active or not, so "platform" column is needed
    - A restaurant can have multiple locations, so coordinates variables are also needed
    - Same coordinates may have different restaurant names due to them being entered into the system during different time period, so "restaurant_chain" column is also needed
- After performing deduplication, about 85,031 entries were deleted as they are deemed repeated

In [604]:

dim(df)
dim(unique(df_final[c("restaurant_chain","platform","longitude","latitude")]))
dim(unique(df_final[c("restaurant_chain","longitude","latitude")]))

dim(df) - dim(unique(df_final[c("restaurant_chain","platform","longitude","latitude")]))


[1] 644089     13

[1] 559058      4

[1] 498516      3

[1] 85031     9

In [614]:
df_final_unique <- distinct(df_final, restaurant_chain,latitude,longitude,.keep_all = TRUE)
df_final <- distinct(df_final, restaurant_chain,platform, latitude,longitude,.keep_all = TRUE)

##### 4.3 Export the cleaned up dataset into a new csv file #####

In [615]:
write.csv(df_final_unique,"df_final_unique_V2.csv", row.names = TRUE)
write.csv(df_final,"df_final_V2.csv", row.names = TRUE)

In [617]:
dim(df_final_unique)
dim(df_final)

[1] 498516     11

[1] 559058     11

In [622]:
head(df_final_unique)

restaurant_chain   branch                   type active platform latitude
1 loco coco          NA                       SMBs TRUE   caviar   40.7650 
2 musubi             NA                       SMBs FALSE  caviar   45.5046 
3 the bakers' lounge NA                       SMBs FALSE  caviar   38.9246 
4 parm               battery park city        SMBs TRUE   caviar   40.7142 
5 the plant cafe     california & pine - fidi SMBs FALSE  caviar   37.7928 
6 minton's           NA                       SMBs FALSE  caviar   40.8047 
  longitude city          state                country delivery_radius
1  -73.9660 new york      new york             US      NA             
2 -122.6438 multnomah     oregon               US      NA             
3  -76.9901 washington    district of columbia US      NA             
4  -74.0156 new york      new york             US      NA             
5 -122.3979 san francisco california           US      NA             
6  -73.9523 new york      new york             US      NA